CCDFDataModel/CReadPNG.cpp

```diff
@@ -162,21 +158,25 @@ CReadPNG::CPNGRaster *CReadPNG::read_png_file(const char *file_name, bool pngRea
   pngRaster->data = new unsigned char[pngRaster->width * pngRaster->height * 4];
   for (size_t y = 0; y < pngRaster->height; y++) {
     unsigned char *line = row_pointers[y];
-    for (size_t x = 0; x < pngRaster->width / pngwidthdivisor; x += 1) {
+    for (size_t x = 0; x < (pngRaster->width / pngwidthdivisor) + 1; x += 1) {
 
-      if (bpp == 1) {
-        unsigned char i = line[x * bpp + 0];
-        if (bit_depth == 1) {
-          int d = 1;
+      /*  Each pixel is a palette index, using PLTE */
+      if (color_type == 3) {
+        unsigned char i = line[x * 1 + 0];
+        if (bit_depth == 1) { // 1 byte is 8 pixels (colormap of 2 colors)
+          int d = 128;
           for (int j = 0; j < 8; j++) {
-            pngRaster->data[x * 32 + j * 4 + 0 + (y * pngRaster->width * 4)] = palette[(i & d) / d].red;
-            pngRaster->data[x * 32 + j * 4 + 1 + (y * pngRaster->width * 4)] = palette[(i & d) / d].green;
-            pngRaster->data[x * 32 + j * 4 + 2 + (y * pngRaster->width * 4)] = palette[(i & d) / d].blue;
-            pngRaster->data[x * 32 + j * 4 + 3 + (y * pngRaster->width * 4)] = 255;
-            d = d + d;
+            int nx = x * 32 + j * 4;
+            if (nx + 3 < pngRaster->width * 4) {
+              pngRaster->data[nx + 0 + (y * pngRaster->width * 4)] = palette[(i & d) / d].red;
+              pngRaster->data[nx + 1 + (y * pngRaster->width * 4)] = palette[(i & d) / d].green;
```

Comment:
You might be able to save some multiplications here, by doing them just once. You could save on additions by using ++.
This is being done for each pixel, so that could add up.